wal/wal.go

```diff
@@ -480,6 +480,106 @@ func (w *WAL) ReadAll() (metadata []byte, state raftpb.HardState, ents []raftpb.
 	return metadata, state, ents, err
 }
 
+// Verify reads through the records of the current WAL and verifies that it is not corrupt.
+// Unlike ReadAll, it does not return the contents of the WAL.
+// Verify cannot be called twice on the same WAL, or after a ReadAll call.
+// If opened in write mode, it must read out all records until EOF. Or an error
+// will be returned.
+// If opened in read mode, it will try to read all records if possible.
+// It is recommended to call Verify only when WAL is opened in read mode.
+// If it cannot read out the expected snap, it will return ErrSnapshotNotFound.
+// If loaded snap doesn't match with the expected one, it will return
+// all the records and error ErrSnapshotMismatch.
+// After Verify, the WAL will be ready for appending new records.
+func (w *WAL) Verify() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var metadata []byte
+	var err error
+
+	rec := &walpb.Record{}
+	decoder := w.decoder
+
+	var match bool
+	for err = decoder.decode(rec); err == nil; err = decoder.decode(rec) {
+		switch rec.Type {
+		case metadataType:
+			if metadata != nil && !bytes.Equal(metadata, rec.Data) {
+				return ErrMetadataConflict
+			}
+			metadata = rec.Data
+		case crcType:
+			crc := decoder.crc.Sum32()
+			// current crc of decoder must match the crc of the record.
+			// do no need to match 0 crc, since the decoder is a new one at this case.
+			if crc != 0 && rec.Validate(crc) != nil {
+				return ErrCRCMismatch
+			}
+			decoder.updateCRC(rec.Crc)
+		case snapshotType:
+			fmt.Println("case snapshotType")
+			var snap walpb.Snapshot
+			pbutil.MustUnmarshal(&snap, rec.Data)
+			if snap.Index == w.start.Index {
+				if snap.Term != w.start.Term {
+					return ErrSnapshotMismatch
+				}
+				match = true
+			}
+		// We omit all entry and state type records as these
+		// are not necessary for validating the WAL contents
+		case entryType:
+		case stateType:
+		default:
+			return fmt.Errorf("unexpected block type %d", rec.Type)
+		}
+	}
+
+	switch w.tail() {
+	case nil:
+		// We do not have to read out all entries in read mode.
+		if err != io.EOF && err != io.ErrUnexpectedEOF {
+			return err
+		}
+	default:
+		// We must read all of the entries if WAL is opened in write mode.
+		if err != io.EOF {
+			return err
+		}
+		if _, err = w.tail().Seek(w.decoder.lastOffset(), io.SeekStart); err != nil {
+			return err
+		}
+		if err = fileutil.ZeroToEnd(w.tail().File); err != nil {
+			return err
+		}
+	}
+
+	err = nil
+	if !match {
+		err = ErrSnapshotNotFound
+	}
+
+	// close decoder, disable reading
+	if w.readClose != nil {
+		w.readClose()
+		w.readClose = nil
+	}
+	w.start = walpb.Snapshot{}
```

Comment:
should verify change anything? it should be a readonly operation, right?

Reply:
You're right. I have re-written `Verify` to be a read-only operation. I was able to achieve this by re-creating the `decoder` within the `Verify` function, so that `w.decoder` is untouched, and can be later used by `ReadAll` as usual. To do this, I have refactored the `openAtIndex` function by moving some code into two helper functions `selectWALFiles` and `openWALFiles`, which I can call in `Verify` in order to create a new `decoder` object.